orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -634,50 +634,70 @@ void TrainingRunner::RunWithUpdate(VectorString& feed_names,
                                    VectorString& fetch_names,
                                    std::vector<MLValue>& feeds,
                                    std::vector<MLValue>& fetches) {
-  // Cyclically pick up a worker ID.
-  const size_t worker_id = step_ % params_.pipeline_parallel_size;
-
-  // Wait for the previous work to finish its job.
-  // Its resource cannot be overrided when it's still working.
-  pipeline_worker_pool_.Join(worker_id);
-
-  // Copy thread-used variable to thread-specific buffer to maintain their life.
-  pipeline_worker_pool_.worker_states[worker_id].feed_names = feed_names;
-  pipeline_worker_pool_.worker_states[worker_id].feeds = feeds;
-  pipeline_worker_pool_.worker_states[worker_id].fetch_names = fetch_names;
-  pipeline_worker_pool_.worker_states[worker_id].fetches = std::vector<MLValue>();
-
-  Status status = Status::OK();
-  pipeline_worker_pool_.workers[worker_id] = std::thread([&](
-      const size_t worker_id, const size_t step) {
+  if (params_.pipeline_parallel_size > 1) {
+    // Cyclically pick up a worker ID.
+    const size_t worker_id = step_ % params_.pipeline_parallel_size;
+
+    // Wait for the previous work to finish its job.
+    // Its resource cannot be overrided when it's still working.
+    pipeline_worker_pool_.Join(worker_id);
+
+    // Copy thread-used variable to thread-specific buffer to maintain their life.
+    pipeline_worker_pool_.worker_states[worker_id].feed_names = feed_names;
+    pipeline_worker_pool_.worker_states[worker_id].feeds = feeds;
+    pipeline_worker_pool_.worker_states[worker_id].fetch_names = fetch_names;
+    pipeline_worker_pool_.worker_states[worker_id].fetches = std::vector<MLValue>();
```

Comment:
Not introduced in this PR, but would it be simpler to capture some or all of these by value in the closure passed to `std::thread()` on L652, instead of copying them into a member variable? (Same question for the similar code in  `RunWithoutUpdate()` below.)

Indeed, since we call `JoinAll()` and then read `worker_states[worker_id].fetches`  on L681, back in the main thread, I'm not sure why these can't be captured by reference in the closure. Perhaps I'm missing a corner case?

Reply:
Sure. For RunWithUpdate(...), we use capture-by-reference for all variables. For RunWithoutUpdate(...), we capture `feed`, `feed_names`, `fetch_names` by value.

---
In reply to: [441846878](https://github.com/microsoft/onnxruntime/pull/4265#discussion_r441846878) [](ancestors = 441846878)